test/framework/src/main/java/org/elasticsearch/transport/nio/NioGroup.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.transport.nio.channel.ChannelFactory;
+import org.elasticsearch.transport.nio.channel.NioServerSocketChannel;
+import org.elasticsearch.transport.nio.channel.NioSocketChannel;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class NioGroup {
+
+    private final Logger logger;
+
+    private final int acceptorCount;
+    private final ThreadFactory acceptorThreadFactory;
+    private final ArrayList<AcceptingSelector> acceptors;
+    private final BiFunction<Logger, Supplier<SocketSelector>, AcceptorEventHandler> acceptorEventHandlerFunction;
+    private RoundRobinSupplier<AcceptingSelector> acceptorSupplier;
```

Comment:
can be final?

Reply:
I don't think they can be final in the current design. Creating a `Selector` can throw an exception which is why I do it in `start()`.

I guess the options are:
1. I could make them `volatile`
2. I could throw `IOException` from the ctor. Although with handling exceptions and closing in the ctor.

Thoughts?

Reply:
I am leaning towards 2. it's much cleaner